clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -613,9 +615,15 @@ private void failBatch(ProducerBatch batch, long baseOffset, long logAppendTime,
             transactionManager.removeInFlightBatch(batch);
             if (adjustSequenceNumbers)
                 transactionManager.adjustSequencesDueToFailedBatch(batch);
+            if (exception instanceof UnknownProducerException)
+                // We don't want to return the UnknownProducerException to the user because we already have the
+                // OutOfOrderSequenceException to indicate data loss, and there is no point having two of them
+                // indicating the same thing.
+                exception = Errors.OUT_OF_ORDER_SEQUENCE_NUMBER.exception();
```

Comment:
I guess another option might be to let `UnknownProducerException` extend from `OutOfOrderSequenceException`. The user would still only need to handle the latter.